lib/oauth2/client.rb

```diff
@@ -143,18 +145,23 @@ def get_token(params, access_token_opts = {}, access_token_class = AccessToken)
         opts[:headers] = {}
       end
       opts[:headers].merge!(headers)
+
       response = request(options[:token_method], token_url, opts)
       response_contains_token = response.parsed.is_a?(Hash) &&
                                 (response.parsed['access_token'] || response.parsed['id_token'])
 
+      if nested_access_token_hash = dig_response(response, options[:dig_nested_keys_for_access_token])
+        response_contains_token = true
```

Comment:
Shouldn't this only set `response_contains_token` to be true if `nested_access_token_hash` actually exists?

Reply:
Can we use this solution to solve the case above which exists to solve the problem that Microsoft implementations only supply `id_token` not `access_token` when only then openid scope is requested. 

That case is a bit more complicated in that it wants `access_token` if it exists otherwise `id_token`. It also exposes the problem that you identity here with is that pulling out the token is smeared across a few classes. Presumably you're doing the merge later so you don't have to subclass AccessToken, reimplementing `from_hash` to get the nested access token.

How about we just generalize the concept with a client option proc called `extract_access_token` by default it would be `Proc.new { |response|  response.parsed['access_token'] }`. Microsoft clients would be `Proc.new { |response|  response.parsed['access_token'] || response.parsed['id_token'] }` and your use case  `Proc.new { |response| response.parsed.dig(:data,:access_token) }`

Reply:
That is an excellent idea @jonspalmer